File: feature_detect/TrackCreator.cpp
#include <iostream>
#include <stdio.h>
#include <unistd.h>
#include <cmath>
#include "FeatureDetectorSIFT.h"
#include "KeypointDescriptor.h"
#include "Image.h"
#include "opencv2/calib3d.hpp"
#include "opencv2/xfeatures2d/nonfree.hpp"
#include "KeypointMetadata.h"
#include "TrackCreator.h"

using namespace cv;
using namespace cv::xfeatures2d;
using namespace std;

namespace SceneComps {
    
void TrackCreator::computeTracks(vector<Image> images, vector<KeypointMetadata> pointList) {
    
    vector<KeyPoint> key1;
    vector<KeyPoint> key2;
    Image image1 = images[0];
    Image image2 = images[1]; 
    vector<DMatch> matches;
    KeypointMetadata point1;
    KeypointMetadata point2;
    int totalmatches = 0;
    int repeats;
    int different;
    
    //For every pair of images, match images
    for (vector<Image>::size_type i = 0; i != images.size(); i++){
        for (vector<Image>::size_type j = i+1; j != images.size(); j++){
            key1.clear();
            key2.clear();
            image1 = images[i];
            image2 = images[j];
            matches.clear();
            
            TrackCreator::computeAndFilterPairMatches(key1, key2, image1, image2, matches);
            
//            matches.clear();
//            matches.push_back(DMatch(0,0,0));
//            matches.push_back(DMatch(1,1,0));
//            matches.push_back(DMatch(2,2,0));
            
            //Add all keypoints to pointlist, if this is our first time seeing them
            if (i==0){
                if (j==1){   
                    for (vector<KeyPoint>::size_type a = 0; a != key1.size(); a++){
                        pointList.push_back(KeypointMetadata(a, i));
                    }
                }
                for (vector<KeyPoint>::size_type b = 0; b != key2.size(); b++){
                    pointList.push_back(KeypointMetadata(b, j));
                }
            }
            
            //for each entry in matches between images i and j
            for (vector<DMatch>::size_type k = 0; k != matches.size(); k++){
                totalmatches ++;
                //look at all the points
                for (vector<KeypointMetadata>::size_type p = 0; p != pointList.size(); p++){
                    // if it has the same image and the same point2d, 
                    //then it's the container for the match's first point
                    if (pointList[p].imageIndex == i &&  pointList[p].pointIndex == matches[k].queryIdx ){
                         //find the container matching the point matched in image j
                        for (vector<KeypointMetadata>::size_type q = 0; q != pointList.size(); q++){  
                            if (pointList[q].imageIndex == j &&  pointList[q].pointIndex == matches[k].trainIdx){
                                pointList[q].matchIndexList.push_back(p);
                                pointList[p].matchIndexList.push_back(q);
                                //cout << "matching edge " << p << " with " << q <<"\n";
                            }
                        }
                    }
                }
            }
            
        }
    }
    findAndFilterSubgraphs(pointList);
    
    
    //print results
    int matched = 0;
    int unmatched = 1;
    for (vector<KeypointMetadata>::size_type p = 0; p != pointList.size(); p++){
        cout << pointList[p].imageIndex << ":" << pointList[p].trackNumber << " ";
//        cout << pointList[p].imageIndex << ":";
//        if (pointList[p].matchIndexList.size() > 0){
//            for (vector<int>::size_type m = 0; m != pointList[p].matchIndexList.size(); m++){
//                cout << pointList[pointList[p].matchIndexList[m]].imageIndex;
//            }
//            cout << " ";
//            matched ++;
//        } else {
//            cout << "_ ";
//            unmatched ++;
//        }
    }
    cout << "\n" << pointList.size() << "\n" <<matched << "matched, " << unmatched << "unmatched \n"; 
    cout << "total matches: " << totalmatches ;
    
} 
    
void TrackCreator::findAndFilterSubgraphs(vector<KeypointMetadata> &pointList){
    //Takes the pointList and figures out the trackNumber and matched of each point
    int currentTrack = 1;
    vector<int> currentTrackIndices;
    vector<int> imagesUsed;
    int repeatedImage = 0;
    //go through pointList
    for (vector<KeypointMetadata>::size_type i = 0; i != pointList.size(); i++){
        currentTrackIndices.clear();
        imagesUsed.clear();
        if (pointList[i].visited==0){
            traverseFromPoint(pointList, currentTrackIndices, i);
            if (currentTrackIndices.size() > 1){
                repeatedImage = 0;
                for (vector<int>::size_type j = 0; j != currentTrackIndices.size(); j++){
                    //if this point's image hasn't been seen in this track
                    if (find(imagesUsed.begin(),imagesUsed.end(),pointList[currentTrackIndices[j]].imageIndex) == imagesUsed.end()){
                        //add the image to list of used images
                        imagesUsed.push_back(pointList[currentTrackIndices[j]].imageIndex);
                    } else {
                        repeatedImage = 1;
                    }
                }
                //none of the currentTrack points used the same image twice
                if (repeatedImage == 0){
                    for (vector<int>::size_type k = 0; k != currentTrackIndices.size(); k++){
                        pointList[currentTrackIndices[k]].trackNumber = currentTrack;
                    }
                    currentTrack ++;
                //there's a repeat image
                } else {
                    for (vector<int>::size_type k = 0; k != currentTrackIndices.size(); k++){
                        pointList[currentTrackIndices[k]].trackNumber = -1;
                    }
                }
            }
        }
    }
}
    
void TrackCreator::traverseFromPoint(vector<KeypointMetadata> &pointList, vector<int> &currentTrackIndices, vector<KeypointMetadata>::size_type index){
    currentTrackIndices.push_back(index);
    pointList[index].visited = 1;
    for (vector<int>::size_type i = 0; i != pointList[index].matchIndexList.size(); i++){
        if (pointList[pointList[index].matchIndexList[i]].visited == 0){
            traverseFromPoint(pointList, currentTrackIndices, pointList[index].matchIndexList[i]);
        }
    }
}
    
void TrackCreator::computeAndFilterPairMatches(vector<KeyPoint> &key1, vector<KeyPoint> &key2, Image image1, Image image2, vector<DMatch> &matches){
    
    FeatureDetectorSIFT siftDetector = FeatureDetectorSIFT();
    vector<KeypointDescriptor> keypoints1 = siftDetector.detect(image1);
    vector<KeypointDescriptor> keypoints2 = siftDetector.detect(image2);

    // Convert descriptors back to cv keypoints :(
    key1.assign(keypoints1.begin(), keypoints1.end());
    key2.assign(keypoints2.begin(), keypoints2.end());
    //vector<KeyPoint> sift_keypoints2 = vector<KeyPoint>(keypoints2.begin(), keypoints2.end());

    //STUFF FROM THE OPEN CV EXAMPLE BELOW
    // https://github.com/npinto/opencv/blob/master/samples/cpp/matcher_simple.cpp
    cv::Ptr<Feature2D> f2d = xfeatures2d::SIFT::create();
    
    Mat descriptors1, descriptors2; 
    f2d->compute(image1.matrix, key1, descriptors1);
    f2d->compute(image2.matrix, key2, descriptors2);
    
    BFMatcher matcher;
    matcher.match(descriptors1, descriptors2, matches);
    
    vector<Point2f> ptList1;
    vector<Point2f> ptList2;
    
    vector<int> queryIdxs;
    vector<int> trainIdxs;
    
    for (vector<DMatch>::size_type i = 0; i != matches.size(); i++){
        queryIdxs.push_back(matches[i].queryIdx);
        trainIdxs.push_back(matches[i].trainIdx);
    }
    
    KeyPoint::convert(key1, ptList1, queryIdxs);
    KeyPoint::convert(key2, ptList2, trainIdxs);
    
    vector<uchar> funOut;
    
    //press 8 for RANSAC
    Mat F = findFundamentalMat(ptList1, ptList2, 8, 3, .99, funOut);
    
    vector<int> funOutInt(funOut.begin(), funOut.end());
    vector<DMatch> filtered_matches;
    
    for (vector<int>::size_type i = 0; i != funOut.size() ; i++){
        if (funOutInt[i]==0){
            filtered_matches.push_back(matches[i]);
        }
    }
    
    int repeatsTrain = 0;
    int repeatsQuery = 0;
    int different = 0;
    
     //for testing repeats
    for (vector<DMatch>::size_type x = 0; x != filtered_matches.size(); x++){
        for (vector<DMatch>::size_type y = x+1; y != filtered_matches.size(); y++){
            if (filtered_matches[x].trainIdx == filtered_matches[y].trainIdx){
                repeatsTrain++;
            }
            else if (filtered_matches[x].queryIdx == filtered_matches[y].queryIdx){
                repeatsQuery++;
            }
            else {
                different++;
            }
        }
    }
            cout << "Qrepeats: " << repeatsQuery <<" | Trepeats: " << repeatsTrain << " | different: "<< different << "\n";
    
    matches.clear();
    matches.assign(filtered_matches.begin(),filtered_matches.end());
    
}    
    
    
}

File: python/track_creator.py

class KeypointMetadata:

    def __init__(self, kp, image):
        self.point_index = kp
        self.image_index = image
        self.visited = 0
        self.track_number = -1
        self.match_index_list = []


class TrackCreator:

    @classmethod
    def computeTracks(cls, images):
        fundamental_mat_array = []
        point_list = []
        key1 = []
        key2 = []

        image1 = images[0]
        image2 = images[1]

        matches = []

        point1 = None
        point2 = None

        total_matches = 0
        repeats = 0
        different = 0

        for i in range(len(images)):
            fundamental_mat_array.append([])
            # for j in range(0,i+1):
            #     fundamental_mat_array[i].
            for j in range(i + 1, len(images)):
                key1 = []
                key2 = []
                image1 = images[i]
                image2 = images[j]
                matches = []

                key1, key2, matches, fundamental_mat = self.computerAndFilterPairMatches(image1, image2)

                if i == 0:
                    if j == 1:
                        for a in range(len(key1)):
                            point_list.append(KeypointMetadata(a, i))
                    for b in range(len(key2)):
                        point_list.append(KeypointMetadata(b, j))

                for k in range(0, len(matches)):
                    total_matches += 1
                    for p in range(0, len(point_list)):
                        if point_list[p].image_index == i and point_list[p].point_index = matches[k].queryIdx:
                            for q in range(0, len(point_list)):
                                if point_list[q].image_index == j and point_list[q].point_index == matches[k].trainIdx:
                                    point_list[q].match_index_list.append(p)
                                    point_list[p].match_index_list.append(q)

            track_list = cls.find_and_filter_subgraphs(point_list)

        # print results
        matched = 0
        unmatched = 1
        for p in range(0, len(point_list)):
            print point_list[p].image_index, ": ", point_list[p].track_number

        print "Does matrix[y][x] exist?"
        for m in range(0, len(fundamental_mat_array)):
            for n in range(0, len(fundamental_mat_array[m])):
                if fundamental_mat_array[m][n].rows == 0:
                    print 0
                else:
                    print 1
            print

    # Finds valid tracks which are connected subgraphs of at least two points that have each image no more than once
    # Updates a referenced pointList of the graph
    @classmethod
    def find_and_filter_subgraphs(cls, point_list):
        current_track = 0
        currrent_track_indicies = []
        images_used = []
        repeated_image = 0

        for i in range(0, len(point_list)):
            currrent_track_indicies = []
            images_used = []
            if point_list[i].visited == 0:
                cls.traverse_from_point(i, point_list, currrent_track_indicies)

                if len(currrent_track_indicies) > 1:
                    repeated_image = 0
                    # if this point's image hasn't been seen in this track
                    for j in range(0, len(currrent_track_indicies)):
                        if currrent_track_indicies[j].imageIndex in images_used:
                            images_used.append(point_list[currrent_track_indicies[j]].image_index)
                        else:
                            repeated_image = 1

                    if repeated_image = 0:
                        track_list.append([])
                        for k in range(0, len(currrent_track_indicies)):
                            point_list[currrent_track_indicies[k]].track_number = current_track
                            track_list[current_track].append(point_list[currrent_track_indicies[k]])

                        current_track += 1

                    else:
                        for k in range(0, len(currrent_track_indicies)):
                            point_list[currrent_track_indicies[k]].track_number = -2

    @classmethod
    def traverse_from_point(cls, index, point_list, currrent_track_indicies):
        currrent_track_indicies.append(index)
        point_list[index].visited = 1
        for i in range(0, len(point_list[index].match_index_list)):
            if point_list[point_list[index].match_index_list[i]].visited == 0:
                cls.traverse_from_point(point_list[index].match_index_list[i], point_list, currrent_track_indicies)

    # returns key1, key2, matches, fundamental_mat
    @classmethod
    def compute_and_filter_pair_matches(cls, image1, image2):
        sift = cv2.xfeatures2d.SIFT_create()
        image1kps, image1descs = sift.detectAndCompute(image1, None)
        image2kps, image2descs = sift.detectAndCompute(image2, None)

        bf = cv2.BFMatcher(cv2.NORM_L1, crossCheck=True)
        matches = bf.match(image1descs, image2descs)

        # query_indexes = []
        # train_indexes = []

        # for i in range(0, len(matches)):
        #     query_indexes.append(matches[i].queryIdx)
        #     train_indexes.append(matches[i].trainIdx)

        # fundamental_mat =

        pts1 = []
        pts2 = []

        for match in matches:
            pts2.append(img2.kps[match.trainIdx].pt)
            pts1.append(img1.kps[match.queryIdx].pt)

        pts1 = numpy.int32(pts1)
        pts2 = numpy.int32(pts2)

        F, mask = cv2.findFundamentalMat(pts1, pts2, cv2.FM_LMEDS)
        # E, mask = cv2.findEssentialMat(pts1, pts2)

        # TODO: do something to filter matches?

        return pts1, pts2, matches, F
